### PR TITLE
fix(eslint): display error if eslint is not found

### LINKED ISSUE #70: `eslint` dependency
Instead of hard code `eslint` directory, use [require.resolve](https://nodejs.org/api/globals.html#globals_require_resolve) to detect it.

If no `eslint` is found in the `PATH` display proper error.

Refer #69

### PR DESCRIPTION
**Pull request by sarbbottam**
anywhere in the path

fixed #70

### CODE REVIEW

**Comment by sarbbottam** [comment]
@ta2edchimp could you take a look?

File: src/lib/rule-finder.js [hunk 1]
```diff
@@ -51,8 +59,13 @@ function _getPluginRules(config) {
...
 function _getAllAvailableRules(pluginRules) {
-  var allRules = fs
-    .readdirSync('./node_modules/eslint/lib/rules')
+  var allRules
+  var eslintPath = require.resolve('eslint')
+  eslintPath = eslintPath.match(/.*eslint/)[0]
+  // no check required for existence of eslint here
+  // refer line#9-line#14
+  allRules = fs
+    .readdirSync(eslintPath)
```
Comment (ta2edchimp) [inline src/lib/rule-finder.js]: Does this really resolve to the *lib/rules* subdir? Rather looks like just eslint's base path.

**Comment by ta2edchimp** [comment]
The problem I see, is in the lib path's resolution, when ESLint is installed globally (we need to get the '/lib/rules' subdir's contents, I don't see that functionality from the code).

The next issue is, when ESLint is installed globally, iirc also plugins and configs have to be global, or is this information deprecated? Depending on this, plugins have to be resolved differently (in order to get their rules). Or am I missing something?

File: src/lib/rule-finder.js (hunk 1 above)
Comment (sarbbottam) [inline src/lib/rule-finder.js]: My bad! I should have added `'/lib/rules'` to the `eslintPath`.

**Comment by sarbbottam** [comment]
>The next issue is, when ESLint is installed globally, iirc also plugins and configs have to be global, or is this information deprecated? Depending on this, plugins have to be resolved differently (in order to get their rules). Or am I missing something?

I see your point, but folks would like to use it as a global utility.
May be, we can warn them if used as global utility?

**Comment by ta2edchimp** [comment]
>May be, we can warn them if used as global utility?

That's a good idea 👍 

But what I really tried to say is, the code currently does not cover `plugins` to be loaded/taken into consideration for all our comparisons, when installed globally (only locally installed). See [`rule-finder.js`](https://github.com/sarbbottam/eslint-find-rules/blob/master/src/lib/rule-finder.js#L35-L52):

```JavaScript
   var pluginConfig = require('eslint-plugin-' + plugin)
```

I think we'd have to `require` these globally too, if ESLint has been found only globally installed.
Do you know what I mean?

**Comment by sarbbottam** [comment]
> I think we'd have to require these globally too, if ESLint has been found only globally installed.

>> May be, we can warn them if used as global utility?

I should have used the word `error` instead of `warn` and exit.
I wanted suggest, that if the utility is being used globally, error out if the corresponding node modules are not found.

---

Then again; Is it possible for a globally installed utility to consume local modules?

**Comment by ta2edchimp** [comment]
>Then again; Is it possible for a globally installed utility to consume local modules?

I guess not (directly), what could also be the reason, a globally installed ESLint does (did?) not find locally installed configs or plugins.  
One could fiddle around using `process.cwd()`, but I guest this would be very error prone or really bloat up the necessary code. Tbh, I'd rather stick to a note in `README`, pointing out that it's considered bad practice to use such things globally ...

**Comment by ta2edchimp** [comment]
To add:
We **could** try to resolve it locally by the simplest possible means and print out that it may be not 100%ly accurate or sth. and to suggest to rather install all the things locally?

**Comment by sarbbottam** [comment]
>I'd rather stick to a note in README, pointing out that it's considered bad practice to use such things globally ...

👍 

And let the utility display error if this is being used as a global utility.

**Comment by ta2edchimp** [comment]
That would be even easier (for us), yeah.

I'd really prefer that way ... an inaccurate output has zero degree of information in such a case, either way.

**Comment by sarbbottam** [comment]
I will update the PR accordingly.

**Comment by sarbbottam** [comment]
This is what I am thinking ...

- use [which](https://www.npmjs.com/package/which) to find out if `eslint-find-rules` is instaled globally
if yes
 - use `__dirname` to find if the current instance is global or not
 if global 
   - display error

Let me know your thoughts.

**Comment by sarbbottam** [comment]
@ta2edchimp does the approach look ok to you?

**Comment by ta2edchimp** [comment]
For ESLint being absent, this LGTM.

But personally, I'd like to catch the cases where it's used being installed globally and give a warning about it (imo, plugins and configs used, *have* to be installed globally, too).

**Comment by sarbbottam** [comment]
> But personally, I'd like to catch the cases where it's used being installed globally and give a warning about it (imo, plugins and configs used, have to be installed globally, too).

I guess it would be better, to let the user know to, this is a local utility.

>- use [which](https://www.npmjs.com/package/which) to find out if `eslint-find-rules` is instaled globally
if yes
 - use `__dirname` to find if the current instance is global or not
 - if global 
   - display error

**Comment by ta2edchimp** [comment]
I agree.

**Comment by sarbbottam** [comment]
@ta2edchimp could you take a look?

**Comment by ta2edchimp** [comment]
LGTM 👍

File: package.json [hunk 2]
```diff
@@ -50,6 +50,7 @@
     "sinon": "^1.17.3",
     "travis-after-all": "^1.4.4",
     "validate-commit-msg": "2.6.1",
+    "which": "^1.2.8",
```
Comment (kentcdodds) [inline package.json]: Just noticed this was added as a dependency in this PR but it's not being used in this PR. Was this a mistake?

Comment (sarbbottam) [inline package.json]: It's being used at https://github.com/sarbbottam/eslint-find-rules/pull/72/files/922dd7e8ce917e9bf24b1662436c6c3f333063d0#diff-7f94494686a3b958a7c65884d555437fR5

Comment (kentcdodds) [inline package.json]: Oh, so this is just adding it as a dependency explicitly to fix an outstanding issue?

Comment (sarbbottam) [inline package.json]: As discussed at https://github.com/sarbbottam/eslint-find-rules/pull/72#issuecomment-217050170
>- use [which](https://www.npmjs.com/package/which) to find out if `eslint-find-rules` is instaled globally
if yes
 - use `__dirname` to find if the current instance is global or not
 - if global 
   - display error

Should it be handled differently?

Comment (sarbbottam) [inline package.json]: Ahh, I see, it should be `dependencies` not `devDependency`.

**Comment by ljharb** [comment]
This is a breaking change released in a patch version, and also doesn't appear to be working correctly. Filed as #83.

**Pull request merged**